src/libstd/sys/wasm/mod.rs

```diff
@@ -102,3 +96,230 @@ pub unsafe fn abort_internal() -> ! {
+}
+
+impl GetEnvSysCall {
+    pub fn perform(key: &OsStr) -> Option<OsString> {
+        let key_buf = &AsInner::as_inner(key).inner;
+        receive_buffer(64, |buffer| {
+            let mut call_record = GetEnvSysCall {
+                key_len: key_buf.len(),
+                key_ptr: key_buf.as_ptr(),
+                value_len: buffer.len(),
+                value_ptr: buffer.as_mut_ptr(),
+                result: !0usize
+            };
+            if unsafe { syscall(SysCallIndex::GetEnv, &mut call_record) } {
+                if call_record.result == !0usize {
+                    Err(())
+                } else {
+                    Ok(call_record.result)
+                }
+            } else {
+                Err(())
+            }
+        }).ok().map(|s| {
+            FromInner::from_inner(Buf { inner: s })
+        })
+    }
+}
+
+#[repr(C)]
+pub struct SetEnvSysCall {
+    key_ptr: *const u8,
+    key_len: usize,
+    value_ptr: *const u8,
+    value_len: usize
+}
+
+impl SetEnvSysCall {
+    pub fn perform(key: &OsStr, value: Option<&OsStr>) {
+        let key_buf = &AsInner::as_inner(key).inner;
+        let value_buf = value.map(|v| &AsInner::as_inner(v).inner);
+        let mut call_record = SetEnvSysCall {
+            key_len: key_buf.len(),
+            key_ptr: key_buf.as_ptr(),
+            value_len: value_buf.map(|v| v.len()).unwrap_or(!0usize),
+            value_ptr: value_buf.map(|v| v.as_ptr()).unwrap_or(ptr::null())
+        };
+        unsafe { syscall(SysCallIndex::SetEnv, &mut call_record); }
+    }
+}
+
+pub enum TimeClock {
+    Monotonic = 0,
+    System = 1,
+}
+
+#[repr(C)]
+pub struct TimeSysCall {
+    clock: usize,
+    secs_hi: usize,
+    secs_lo: usize,
+    nanos: usize
+}
+
+impl TimeSysCall {
+    pub fn perform(clock: TimeClock) -> Duration {
+        let mut call_record = TimeSysCall {
+            clock: clock as usize,
+            secs_hi: 0,
+            secs_lo: 0,
+            nanos: 0
+        };
+        if unsafe { syscall(SysCallIndex::Time, &mut call_record) } {
+            Duration::new(
+                ((call_record.secs_hi as u64) << 32) | (call_record.secs_lo as u64),
+                call_record.nanos as u32
+            )
+        } else {
+            panic!("Time system call is not implemented by WebAssembly host");
+        }
+    }
+}
+
+unsafe fn syscall<T>(index: SysCallIndex, data: &mut T) -> bool {
+    __rust_wasm_syscall(index as usize, data as *mut T as *mut Void)
+}
+
+#[cfg(feature = "wasm_syscall")]
+#[allow(private_no_mangle_fns)]
+#[no_mangle]
+#[inline(never)]
+unsafe fn __rust_wasm_syscall(index: usize, data: *mut Void) -> bool {
+    extern {
+        fn rust_wasm_syscall(index: usize, data: *mut Void) -> usize;
+    }
+    rust_wasm_syscall(index, data) != 0
+}
+
+#[cfg(not(feature = "wasm_syscall"))]
+#[allow(private_no_mangle_fns)]
+#[no_mangle]
+#[inline(never)]
+unsafe fn __rust_wasm_syscall(_index: usize, _data: *mut Void) -> bool {
```

Comment:
Er what happened here? It seems like this sort of construction will prevent optimizing the implementations when the syscalls are turned off?

Reply:
@alexcrichton If you scroll up you can see one of my comments where I've requested this. This is precisely the intent. (There really should have been a comment here which explains it.) If the function is non-inlineable and has a unique name I can easily replace it in `cargo-web` when postprocessing the `.wasm` without having to recompile the whole `std`. (So it will make my life *significantly* easier. *Of course* I have zero expectations this being stable in any way, shape or form.)

Is something like this unacceptable to you? ):

Although from what I know this isn't enough to prevent the function from being inlined, which is why my original comment had an empty `asm!` in it. (cc @Diggsey) Consider this program:

```rust
#![feature(asm)]

#[allow(private_no_mangle_fns)]
#[no_mangle]
#[inline(never)]
unsafe fn foo() -> bool {
    false
}

#[allow(private_no_mangle_fns)]
#[no_mangle]
#[inline(never)]
unsafe fn bar() -> bool {
    asm!("");
    false
}

#[no_mangle]
pub extern fn foobar() {
    unsafe {
        foo();
        bar();
    }
}
```

If you compile it with `RUSTFLAGS="-C debuginfo=2" cargo build --target=wasm32-unknown-unknown --release` you get `foobar` looking like this:

```rust
(func $foobar (type 0)
    call $bar)
```

Without the `asm!` the function gets inlined anyway.

Reply:
@koute that's what the custom feature for libstd is, right? I would prefer we *don't* commit to something like this yet in the mainline target.

Reply:
@alexcrichton the more unusable you make the wasm target, the more people are forced to rely on implementations details like this.

All this PR changes are undocumented aspects of our codegen for an unsupported target which do not affect the behaviour of the generated code in any way. If that's now a commitment, then you've already committed to everything that `stdweb` relies on to do its transformation by simply implementing this target in the first place.

Reply:
> @koute that's what the custom feature for libstd is, right? I would prefer we don't commit to something like this yet in the mainline target.

No one is asking for any commitment! This is meant to be an entirely unstable, *temporary* mechanism to hook into the `std`. I'd gladly switch to using the custom feature once it's officially supported by `cargo`. *Right now* both of those approaches are, in a way, hacks, while having to recompile the whole `std` is way more fiddly and will require a lot more effort to support - effort which will most likely go to waste once `cargo` starts to officially support custom `std`s.